raven/utilities/forecasting.py

```diff
@@ -173,3 +182,98 @@ def perform_climatology_esp(model_name, forecast_date, forecast_duration, **kwds
     qsims["member"] = (["member"], avail_years)
 
     return qsims
+
+
+def get_recent_ECCC_forecast(region, climate_model='GEPS'):
+    '''
+    This function generates a forecast dataset that can be used to run raven. 
+    Data comes from the ECCC datamart and collected daily. It is aggregated 
+    such that each file contains forecast data for a single day, but for all 
+    forecast timesteps and all members.
+    
+    The code takes the region shapefile and the climate_model to use, here GEPS
+    by default, but eventually could be GEPS, GDPS, REPS or RDPS.
+    '''
+   
+    # Get the file locations and filenames as a function of the climate model and date
+    if climate_model is 'GEPS':
+        
+        # Forecasts will always be recent so start with today!
+        date_now=dt.datetime.now()
+
+        # Find the most recent data on the server by looping from today and go
+        #backwards 5 days. If no data is present, not much value in forecasting.
+        filein=None
+        for i in range(0,7):
+            
+            # Write the string
+            date=date_now-dt.timedelta(days=i)
+            datefcst=dt.datetime.strftime(date,"%Y%m%d")
+            
+            # try to open the files. If it works, great, if no
+            try:
+                filein='https://pavics.ouranos.ca/twitcher/ows/proxy/thredds/dodsC/birdhouse/testdata/geps_forecast/CMC_geps-raw_latlon0p5x0p5_' + datefcst + '00_allP_allmbrs.nc'
+                ds=xr.open_dataset(filein)
+                # Checking that these exist. IF the files are still processing, possible that one or both are not available!
+                assert 'pr' in ds
+                assert 'tas' in ds
+                break
+            except:
+                filein=None
+        
+        # If we have not found any GEPS data, raise error.
+        if filein is None:
+            raise RuntimeError('Unable to locate valid forecast data up to 5 days prior to today.')
+
+
+        
+        # Here we also extract the times at 6-hour intervals as Raven must have 
+        #constant timesteps.
+        times=[]
+        date = dt.datetime(date.year, date.month, date.day)
+        for single_time in (date + dt.timedelta(hours=n) for n in range(0,384,6)):
+            times.append(single_time)
+
+    elif climate_model is 'GDPS':
+        filein='https://pavics.ouranos.ca/twitcher/ows/proxy/thredds/dodsC/birdhouse/caspar/daily/GDPS_' + dt.datetime.strftime(date,'%Y%m%d') + '.nc'
+
+    elif climate_model is 'RDPS':
+        filein='https://pavics.ouranos.ca/twitcher/ows/proxy/thredds/dodsC/birdhouse/caspar/daily/REPS_' + dt.datetime.strftime(date,'%Y%m%d') + '.nc'
+
+    elif climate_model is 'REPS':
+        filein='https://pavics.ouranos.ca/twitcher/ows/proxy/thredds/dodsC/birdhouse/caspar/daily/RDPS_' + dt.datetime.strftime(date,'%Y%m%d') + '.nc'
```

Comment:
Use string template instead of `if`.

Reply:
@huard @richardarsenault I just realized that the control flow is a bit weird here:

https://github.com/Ouranosinc/raven/blob/wps_geomet/raven/utilities/forecasting.py#L237-L244

as `filein` is not being used further, with models `GDPS`, `RDPS` or `REPS` (the last 3 `elif` clauses).

Shouldn't the creation of `ds`:

https://github.com/Ouranosinc/raven/blob/wps_geomet/raven/utilities/forecasting.py#L216

be pushed farther, i.e. at least after line 245:

https://github.com/Ouranosinc/raven/blob/wps_geomet/raven/utilities/forecasting.py#L245

Reply:
Perhaps, yes! This was done at first because I had to cycle through the previous files to find an existing one. Now, we have the "latest" file which should ne used instead, so the opening of fileIn and checking of the presence of "tas" and "pr" in the file could go after this block.